### ISSUE TITLE
Source in ACL names does not work

### ISSUE THREAD

**Issue by f1-outsourcing** [3proxy/3proxy#740]
It is stated that names in the source of acl's should work.

Source in ACL supports names (resolved on configuration reading to IP) and ranges, but if you want to authenticate based on DNS name you should consider

auth dnsname

as a more reliable mechanism.

_Originally posted by @z3APA3A in https://github.com/3proxy/3proxy/issues/736#issuecomment-1031422793_

This allows me to connect to a website:
```
auth iponly
deny * * 127.0.0.1
#allow * test.local * 80,443
allow * 192.168.220.111 * 80,443
```

This refuses me the connection to the same website:
```
auth iponly
deny * * 127.0.0.1
allow * test.local * 80,443
#allow * 192.168.220.111 * 80,443
```

**Comment by f1-outsourcing** [comment]
@z3APA3A 

I am willing to contribute financially for this change

**Comment by z3APA3A** [comment]
check `auth` `dnsname` in https://3proxy.ru/doc/man3/3proxy.cfg.3.html, probably that's what you need. With `auth dnsname` reverse name is considered as username, FCrDNS is required. Use caching by ip to avoid DNS resolution for every request, e.g.

```
authcache ip  60
auth cache dnsname
allow test.local * * 80,443

```

**Comment by f1-outsourcing** [comment]
> 

No, I do not like using this, that would mean I have to dynamically update PTR records, which I currently can not do. I would like to use just source + username.

eg.
`allow testuser test.local * 80,443`

This is for most dynamic solutions sufficient.

Why not add this `authcache ip  60` and resolving as a default to the source acl?

**Comment by f1-outsourcing** [comment]
Please reopen this issue as 'auth cache dnsname' is not the same as 'names in the source of acl's'

**Comment by z3APA3A** [comment]
I will check if it's hard to implement name resolution for src address.

**Comment by f1-outsourcing** [comment]
Cool. Thanks!!!! :)

**Comment by z3APA3A** [comment]
It's absolutely different, because in target name is taken from the request, e.g. if client does `CONNECT example.com HTTP/1.0` it matches `example.com` in destination ACL. If client does `CONNECT 93.184.216.34 HTTP/1.0` it does not. For source, the is no name, it should be resolved somehow. Either IP address of the client must be resolved via PTR (implemented in auth dnsname) or _every_ name in ACL must be dynamically resolved to IP (on every request?) as you request, it does not sound good for me (despite the fact DNS names are cached) and require separate logic anyway. It's relative simple to make resolution on configuration reload, but if IP changes, you will not see it until configuration is reloaded.

**Comment by f1-outsourcing** [comment]
Maybe not do this at the request level then you indeed only have the ip address and still have to get something via PTR. 

What about upon reading of the config file, you resolve the hostnames in the config to ip addresses. 
Process requests like this until the set dns cache time out
When the dns cache expires you reload/re-resolve these domains in source in the config.
etc.

This skips using PTR's

**Comment by z3APA3A** [comment]
Config is reloaded only on signal or `monitor` event. Cache timeout is checked on name resolution. It means on every request (because request is the only event you actually have) you have to resolve all names in ACL, because you can not know which entry needs to be resolved until you resolve it to IP.  Cached names are resolved from the cache. In the case cache entry is expired due to TTL, name is requested from nserver and cached again. In the worst situation, If you have 100 names in ACL and most entries are expired from cache, it will take 100 DNS requests and a lot of time to process a single request.

**Comment by f1-outsourcing** [comment]
Hmmm, indeed good point, was not really thinking about having quite large amount of source acl hostnames that always need to be checked all at once at dns expire. But I think this 100DNS requests is not likely to occur often. 


What about splitting this into two.

1. Just offering to allow hostnames in src and resolve them at startup/reload
This is also convenient for people that use static ip's. It also gives a very basic support for simple dynamic solutions, by just issuing a reload.

2. lookup at incoming request 
In my case I am assigning a dynamic ip once, and use it until restart. So there is quite a lot of time between ip changes, months even. I think in most cases dynamic ip changes > one day. 
So lets assume this for now.

updating the source acl tuples in cache can be something like: first checking the cache for the incoming ip, get the stored hostname, lookup that hostname and verify if the ip is still the same, if so update new timeout value. This would be your most often occurring situation, just one dns lookup.

We can also expand the source acl tuple to something like (hostname, ip, updates)
So we find hosts that are updating their ip more frequent than others (container/micro services environment)

Say I would put 3proxy source acl dns cache time out on 5 minutes and my hostnames are changing ip every few days. I would be hitting the dns cache lookup 99% of the time. 
And then if I reboot something after a few days 1 request takes a bit longer because of the unknown ip address that needs to be checked against all expired source acl entries.

I think that is quite ok compared to an alternatives, as to manually reload the config, automating some sort of remote rest api request to reload 3proxy config when some container is re-deployed. Have to configure container deployments to issue rest request etc. etc.

It is also possible to start logging warnings of delays to inform users if things get out of hand, so they can rethink their config.

**Comment by f1-outsourcing** [comment]
@z3APA3A Vladimir, what do you think?

**Comment by f1-outsourcing** [comment]
@z3APA3A  what about this feature on startup? It would make my config file a lot more readable.

**Comment by f1-outsourcing** [comment]
@z3APA3A anything new on this?

**Comment by f1-outsourcing** [comment]
@z3APA3A Is it possible to add at least this hostname lookups on loading the acl configuration?

**Comment by f1-outsourcing** [comment]
@z3APA3A How is it with these hostname lookups? That would be very nice to have.

**Comment by f1-outsourcing** [comment]
currently converting the config file with this

```
#!/bin/bash
#

version=1.0
if [ $# -lt 1 ]
then
 echo ""
 echo "convert hostnames into ip addresses (v$version)"
 echo "usage: $0 configfile"
 echo ""
 echo ""
exit;
fi

debug=0
converttld=('marathon.mesos' 'local')
declare -A foundhosts

function gethosts {
  local file="$1"
  local hn=""

  for tld in "${converttld[@]}"
  do
    for hn in $(egrep -o "[0-9a-z\.]+\.$tld" $file)
    do 
      foundhosts[$hn]=""
      [ $debug -eq 1 ] && echo "found hn: $hn"
    done
  done
}

function getips {
  # using getent, skipping installing bind-tools
  local key=""
  local ips=""

  for key in "${!foundhosts[@]}"
  do
     ips=$(getent ahosts "$key"| awk '{print $1}'| sort -u| tr '\n' ',')
     if [ ${#ips} -ge 7 ]
     then
       ips=${ips:0:-1}
       foundhosts[$key]="$ips"
       [ $debug -eq 1 ] && echo "found $key: $ips"
     else
       >&2 echo "could not resolve $key"
     fi
  done
}

function replacehosts {
  local file="$1"
  local key=""

  for key in "${!foundhosts[@]}"
  do
    if [ ${#foundhosts[$key]} -ge 7 ]
    then 
      sed -i "s/$key/${foundhosts[$key]}/g" $file
    else
      echo "deleting config line $key"
      sed -i "/$key/d" $file
    fi
  done
}

gethosts "$1"
getips
replacehosts "$1"
```

**Comment by z3APA3A** [comment]
Check master version, 13979b5df43356818cf64005abe505c23ef05113 should allow hostnames

**Comment by f1-outsourcing** [comment]
I just tried, but is there some syntax change for entering host names

```
test2:/etc/3proxy# 3proxy /etc/3proxy/3proxy.cfg
Invalid IP, IP range or CIDR, line 30
Unable to parse ACL entry, line 30
Command: 'allow' failed with code 1, line 30
```

Line 30 is something like this

`allow * xxxxx.local a.a.a.a/24,b.b.b.b./24`

**Comment by f1-outsourcing** [comment]
I compiled manually from source (git clone --branch master https://github.com/3proxy/3proxy.git) this config runs fine

```
/usr/local/3proxy/conf # cat 3proxy.cfg
nscache 65536
nserver 8.8.8.8
nserver 8.8.4.4

log /logs/3proxy-%y%m%d.log D
rotate 60

auth strong
deny * * 127.0.0.1
allow * 10.10.10.10 127.0.0.1
#allow * www.adobe.com 127.0.0.1
proxy -n
socks
flush
allow admin

admin -p8080
```

and this 

```
/usr/local/3proxy/conf # cat 3proxy.cfg
nscache 65536
nserver 8.8.8.8
nserver 8.8.4.4

log /logs/3proxy-%y%m%d.log D
rotate 60

auth strong
deny * * 127.0.0.1
allow * 10.10.10.10 127.0.0.1
allow * www.adobe.com 127.0.0.1
proxy -n
socks
flush
allow admin

admin -p8080
```

give the error

```
/usr/local/3proxy/conf # 3proxy 3proxy.cfg
Invalid IP, IP range or CIDR, line 11
Unable to parse ACL entry, line 11
Command: 'allow' failed with code 1, line 11

```
